src/libraries/Microsoft.PowerFx.Interpreter/EvalVisitor.cs

```diff
@@ -265,6 +265,100 @@ private async Task<FormulaValue> TryHandleSetProperty(CallNode node, EvalVisitor
 
             return result;
         }
+
+        // Given a TexlFunction, get the implementation to invoke. 
+        private IFunctionInvoker GetInvoker(TexlFunction func)
+        {
+            if (func is IFunctionInvoker invoker)
+            {
+                return invoker;
+            }
+
+            if (func is UserDefinedFunction userDefinedFunc)
+            {
+                return new UserDefinedFunctionAdapter(userDefinedFunc);
+            }
+
+            if (FunctionImplementations.TryGetValue(func, out AsyncFunctionPtr ptr))
+            {
+                return new AsyncFunctionPtrAdapter(ptr);
+            }
+
+            return null;
+        }
+
+        // Adapter for AsyncFunctionPtr to common invoker interface.
+        private class AsyncFunctionPtrAdapter : IFunctionInvoker

```

Comment:
>AsyncFunctionPtrAdapter [](http://example.com/codeflow?start=22&length=23)

nit: can we separate these new classes into a new file?

Reply:
these are nested private classes - which need that to accessing private state on this class. And only used within this file.  
We could consider making a partial class and split to other files.